src/EditorFeatures/Core/Tagging/EfficientTagger.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Classification;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging;
+
+/// <summary>
+/// Root type of all roslyn <see cref="ITagger{T}"/> implementations.  This adds specialized hooks that allow for tags
+/// to be added to a preexisting list, rather than generating a fresh list instance.  This can help with avoiding
+/// garbage by allowing us to pool intermediary lists (especially as some high-level aggregating taggers defer to
+/// intermediary taggers).
+/// </summary>
+internal abstract class EfficientTagger<TTag> : ITagger<TTag>, IDisposable where TTag : ITag
+{
+    /// <summary>
+    /// Produce the set of tags with the requested <paramref name="spans"/>, adding those tags to <paramref name="tags"/>.
+    /// </summary>
+    public abstract void AddTags(NormalizedSnapshotSpanCollection spans, SegmentedList<ITagSpan<TTag>> tags);
+
+    public abstract void Dispose();
+
+    /// <summary>
+    /// Default impl of the core <see cref="ITagger{T}"/> interface.  Forces an allocation.
+    /// </summary>
+    public IEnumerable<ITagSpan<TTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+    {
+        var pooledTags = Classifier.GetPooledList<ITagSpan<TTag>>(out var tags);
```

Comment:
> var pooledTags = Classifier.GetPooledList<ITagSpan<TTag>>(out var tags); [](http://example.com/codeflow?start=7&length=73)

Trying to piece this together, is this method expected to get called?

Reply:
I think I see now, this is the one place that does the bridge between GetTags -> AddTags, and thus needs an allocation

Reply:
Finally, just so I'm clear here, the only real advantage to using the pool here is that if there weren't any tags, it can return an empty array and add the empty one it allocated to the pool?

Reply:
Correct on all counts.

Reply:
Note, this is used for diagnostic tags where the common case is no tags.